src/prof_data.c

```diff
@@ -939,13 +944,82 @@ prof_open_maps(const char *format, ...) {
 static bool
 prof_dump_maps(bool propagate_err) {
 	bool ret;
-	int mfd;
+	int mfd = -1;
 
 	cassert(config_prof);
 #ifdef __FreeBSD__
-	mfd = prof_open_maps("/proc/curproc/map");
-#elif defined(_WIN32)
-	mfd = -1; // Not implemented
+	int mib[4];
+	size_t sz;
+	char *buf;
+
+	mib[0] = CTL_KERN;
+	mib[1] = KERN_PROC;
+	mib[2] = KERN_PROC_VMMAP;
+	mib[3] = prof_getpid();
+	sz = 0;
+
+	if (prof_dump_write(propagate_err, "\nMAPPED_LIBRARIES:\n") &&
+	    propagate_err) {
+		ret = true;
+		goto label_return;
+	}
+
+	if (sysctl(mib, 4, NULL, &sz, NULL, 0) == -1) {
+		ret = true;
+		goto label_return;
+	}
+
+	/* Size for kinfo_vmentry */
+	sz = sz * 4 / 3;
+	buf = mmap(NULL, sz, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANON, -1, 0);
+
+	if (buf == MAP_FAILED) {
+		ret = true;
+		goto label_return;
+	}
+
+	if (sysctl(mib, 4, buf, &sz, NULL, 0) == -1) {
+		ret = true;
+		munmap(buf, sz);
+		goto label_return;
+	}
+
+	char *low = buf;
+	char *high = buf + sz;
+
+	while (low < high) {
+		/* paths + enough room for address ranges and permissions */
+		char map[PATH_MAX+68];
+		struct kinfo_vmentry *e = (struct kinfo_vmentry *)low;
+		size_t esz = e->kve_structsize;
+
+		if (esz == 0) {
+			ret = true;
+			munmap(buf, sz);
+			goto label_return;
+		}
+
+		if (e->kve_path[0] == '\0') {
+			goto loop_cond;
+		}
+
+		malloc_snprintf(map, sizeof(map), "%.12lu-%.12lu %c%c%c %s\n",
```

Comment:
Is 12 right? This must be at least nominally enough for the whole address space, right?

Reply:
Yes it takes not more than 12 large.

Reply:
How come? E.g. if we have a 48-bit address space, it seems we could have as many as ceil(log(2^48)) == 15 digits.